src/views/text_input.rs

```diff
@@ -59,872 +67,545 @@ prop_extractor! {
     }
 }
 
-/// Text Input View
-pub struct TextInput {
-    data: ViewData,
-    buffer: RwSignal<String>,
-    pub(crate) placeholder_text: Option<String>,
-    placeholder_buff: Option<TextLayout>,
-    placeholder_style: PlaceholderStyle,
-    selection_style: SelectionStyle,
-    // Where are we in the main buffer
-    cursor_glyph_idx: usize,
-    // This can be retrieved from the glyph, but we store it for efficiency
-    cursor_x: f64,
-    text_buf: Option<TextLayout>,
-    text_node: Option<NodeId>,
-    // Shown when the width exceeds node width for single line input
-    clipped_text: Option<String>,
-    // Glyph index from which we started clipping
-    clip_start_idx: usize,
-    // This can be retrieved from the clip start glyph, but we store it for efficiency
-    clip_start_x: f64,
-    clip_txt_buf: Option<TextLayout>,
-    // When the visible range changes, we also may need to have a small offset depending on the direction we moved.
-    // This makes sure character under the cursor is always fully visible and correctly aligned,
-    // and may cause the last character in the opposite direction to be "cut"
-    clip_offset_x: f64,
-    selection: Option<Range<usize>>,
-    width: f32,
-    height: f32,
-    // Approx max size of a glyph, given the current font weight & size.
-    glyph_max_size: Size,
-    style: Extracter,
-    font: FontProps,
-    cursor_width: f64, // TODO: make this configurable
-    is_focused: bool,
-    last_cursor_action_on: Instant,
-}
+//TODO add to styles
+static IME_FOREGROUND: Color = Color::BLACK;
 
-#[derive(Clone, Copy, Debug)]
-pub enum Movement {
-    Glyph,
-    Word,
-    Line,
-}
+pub fn text_input(text: impl Fn() -> String + 'static) -> TextInput {
+    let id = Id::next();
 
-#[derive(Clone, Copy, Debug)]
-pub enum Direction {
-    Left,
-    Right,
-}
+    let text_editor = text_editor(text());
+    let editor = create_rw_signal(text_editor.editor().to_owned());
 
-/// Text Input View
-pub fn text_input(buffer: RwSignal<String>) -> TextInput {
-    let id = Id::next();
+    let doc = editor.with_untracked(|e| e.doc_signal());
+    let cursor = editor.with_untracked(|e| e.cursor);
+    let window_origin = create_rw_signal(Point::ZERO);
+    let cursor_line = create_rw_signal(Line::new(Point::ZERO, Point::ZERO));
     let is_focused = create_rw_signal(false);
+    let hide_cursor = editor.with_untracked(|e| e.cursor_info.hidden);
+    let skip_content_sync = create_rw_signal(true);
+    let pointer_focus = create_rw_signal(false);
+
+    // Syncs with external changes to the text. This should *only* happen if the closure
+    // re-runs due to an external update.
+    // For example the closure is subscribed to a signal and the signal is updated by another view.
+    // NOTE: The updates are dispatched in the same order they occur, so this is safe to do
```

Comment:
I don't think the comment past the first line is really needed, that being what signals do & are for.

Reply:
I do want  to keep the note on update order, but I can remove the 2nd line